Mathlib/RingTheory/TwoSidedIdeal/Lattice.lean

```diff
@@ -0,0 +1,113 @@
+      {x | ∃ y ∈ I, ∃ z ∈ J, y + z = x}
+      ⟨0, ⟨zero_mem _, ⟨0, ⟨zero_mem _, zero_add _⟩⟩⟩⟩
+      (by rintro _ _ ⟨x, ⟨hx, ⟨y, ⟨hy, rfl⟩⟩⟩⟩ ⟨a, ⟨ha, ⟨b, ⟨hb, rfl⟩⟩⟩⟩;
+          exact ⟨x + a, ⟨add_mem _ hx ha, ⟨y + b, ⟨add_mem _ hy hb, by abel⟩⟩⟩⟩)
+      (by rintro _ ⟨x, ⟨hx, ⟨y, ⟨hy, rfl⟩⟩⟩⟩
+          exact ⟨-x, ⟨neg_mem _ hx, ⟨-y, ⟨neg_mem _ hy, by abel⟩⟩⟩⟩)
+      (by rintro r _ ⟨x, ⟨hx, ⟨y, ⟨hy, rfl⟩⟩⟩⟩
+          exact ⟨_, ⟨mul_mem_left _ _ _ hx, ⟨_, ⟨mul_mem_left _ _ _ hy, mul_add _ _ _ |>.symm⟩⟩⟩⟩)
+      (by rintro r _ ⟨x, ⟨hx, ⟨y, ⟨hy, rfl⟩⟩⟩⟩
+          exact ⟨_, ⟨mul_mem_right _ _ _ hx, ⟨_, ⟨mul_mem_right _ _ _ hy, add_mul _ _ _ |>.symm⟩⟩⟩⟩)
+
+    suffices (I.ringCon ⊔ J.ringCon) ≤ s.ringCon by
+      intro h; convert this h; rw [rel_iff, sub_zero, mem_mk']; rfl
+
+    refine sup_le (fun x y h => ?_) (fun x y h => ?_) <;> rw [rel_iff] at h ⊢ <;> rw [mem_mk']
+    exacts [⟨_, ⟨h, ⟨0, ⟨zero_mem _, add_zero _⟩⟩⟩⟩, ⟨0, ⟨zero_mem _, ⟨_, ⟨h, zero_add _⟩⟩⟩⟩]
+
+  · rintro ⟨y, ⟨hy, ⟨z, ⟨hz, rfl⟩⟩⟩⟩; exact add_mem _ (mem_sup_left hy) (mem_sup_right hz)
+
+end sup
+
+instance : Inf (TwoSidedIdeal R) where
+  inf := fun I J => { ringCon := I.ringCon ⊓ J.ringCon }
+
+instance : SemilatticeInf (TwoSidedIdeal R) where
+  inf_le_left I J := by rw [ringCon_le_iff]; exact inf_le_left
+  inf_le_right I J := by rw [ringCon_le_iff]; exact inf_le_right
+  le_inf I J K h1 h2 := by rw [ringCon_le_iff] at h1 h2 ⊢; exact le_inf h1 h2
+
+lemma mem_inf {I J : TwoSidedIdeal R} {x : R} :
+    x ∈ I ⊓ J ↔ x ∈ I ∧ x ∈ J :=
+  Iff.rfl
+
+instance : SupSet (TwoSidedIdeal R) where
+  sSup s := { ringCon := sSup $ TwoSidedIdeal.ringCon '' s }
+
+instance : CompleteSemilatticeSup (TwoSidedIdeal R) where
+  sSup_le s I h := by simp_rw [ringCon_le_iff] at h ⊢; exact sSup_le $ by aesop
+  le_sSup s I hI := by rw [ringCon_le_iff]; exact le_sSup $ by aesop
+
+instance : InfSet (TwoSidedIdeal R) where
+  sInf s := { ringCon := sInf $ TwoSidedIdeal.ringCon '' s }
+
+instance : CompleteSemilatticeInf (TwoSidedIdeal R) where
+  le_sInf s I h := by simp_rw [ringCon_le_iff] at h ⊢; exact le_sInf $ by aesop
+  sInf_le s I hI := by rw [ringCon_le_iff]; exact sInf_le $ by aesop
+
+lemma mem_iInf {ι : Type*} {I : ι → TwoSidedIdeal R} {x : R} :
+    x ∈ iInf I ↔ ∀ i, x ∈ I i :=
+  show (∀ _, _) ↔ _ by simp [mem_iff]
+
+lemma mem_sInf {S : Set (TwoSidedIdeal R)} {x : R} :
+    x ∈ sInf S ↔ ∀ I ∈ S, x ∈ I :=
+  show (∀ _, _) ↔ _ by simp [mem_iff]
+
+instance : Top (TwoSidedIdeal R) where
+  top := { ringCon := ⊤ }
+
+instance : Bot (TwoSidedIdeal R) where
+  bot := { ringCon := ⊥ }
+
+lemma mem_bot {x : R} : x ∈ (⊥ : TwoSidedIdeal R) ↔ x = 0 :=
+  Iff.rfl
+
+instance : CompleteLattice (TwoSidedIdeal R) where
+  __ := (inferInstance : SemilatticeSup (TwoSidedIdeal R))
+  __ := (inferInstance : SemilatticeInf (TwoSidedIdeal R))
+  __ := (inferInstance : CompleteSemilatticeSup (TwoSidedIdeal R))
+  __ := (inferInstance : CompleteSemilatticeInf (TwoSidedIdeal R))
```

Comment:
I suspect you can drop these and Lean works it out automatically?

Reply:
unfortunately not. Lean complains
```
fields missing: 'inf_le_left', 'inf_le_right', 'le_inf', 'le_sSup', 'sSup_le', 'sInf_le', 'le_sInf'
```